### PR TITLE
QUA-460: support ESLint 8

### PR DESCRIPTION
**Pull request by dantevvp**
## Purpose

This PR adds support for ESLint 8. Most of the changes introduced in the ESLint 8 update can be seen here:

- [Changes in ESLint 8](https://github.com/eslint/eslint/blob/main/docs/user-guide/migrating-to-8.0.0.md)
- [Changes to the options object](https://github.com/eslint/eslint/blob/main/lib/eslint/eslint.js#L191)

## Description

- Upgraded version of ESLint to `8.8.0`.
- Adapted the code in `lib/eslint.js` to work for the new asynchronous methods `lintFiles` and `isPathIgnored`, replacing the previous synchronous methods `executeOnFiles` and `isPathIgnored`.
- Because of the previous change, `eslint.run()` is now asynchronous.
- Adapted the code in the ESLint 6 patch to import the ConfigArray and ConfigArrayFactory from `@eslint/eslintrc`, which is where those classes live now.
- Adapted integration tests to work with the new asynchronous flow.
- Fixed some errors in system tests.
- Added library `chai-as-promised`, useful for testing asynchronous flows in the app. (In this case, it's being used to test the `eslint.run()` method, given that it's asynchronous now.

## Related issues

https://codeclimate.atlassian.net/browse/QUA-460

### CODE REVIEW

**Review by mileslane — COMMENTED** [review]
LGTM

File: system-tests/complex-config/.codeclimate.yml [hunk 1]
```diff
@@ -4,4 +4,5 @@ plugins:
     enabled: true
     config:
       config: "eslintrc.json"
-    include_paths: [ "/usr/src/app/integration/extends_airbnb/index.js" ]
+    include_paths: 
+      - "/usr/src/app/integration/extends_airbnb/index.js"
```
Comment (fede-moya) [inline system-tests/complex-config/.codeclimate.yml:8]: does this really works now ??, I see some related changes at `inclusionBasedFileListBuilder` on `lib/eslint.js` but not seeing how is it that now we support this other format.

File: lib/eslint.js [hunk 2]
```diff
@@ -123,35 +134,37 @@ function run(console, runOptions) {
...
+      }))
 
       return analysisFiles
     }
   }
 
   function overrideOptions(userConfig) {
     if (userConfig.configPath) {
-      options.configFile = codeDir + "/" + userConfig.configPath
+      options.overrideConfigFile = codeDir + "/" + userConfig.configPath
```
Comment (fede-moya) [inline lib/eslint.js:167]: is this intentional ?

File: integration/eslint_test.js [hunk 3]
```diff
@@ -1,7 +1,9 @@
 const sinon = require("sinon")
-const expect = require("chai").expect
+const chai = require("chai")
+const expect = chai.expect
+chai.use(require('chai-as-promised'))
```
Comment (fede-moya) [inline integration/eslint_test.js:4]: wdyt about adding a note on the  pr description about this new project dependency and an example of why it's handy for testing ?

File: lib/eslint.js [hunk 4]
```diff
@@ -11,10 +11,10 @@ const EngineConfig = require("./engine_config")
 const checks = require("./checks")
 const computeFingerprint = require("./compute_fingerprint")
 
-const CLIEngine = patch.eslint.CLIEngine
-const options = { extensions: [".js"], ignore: true, reset: false, useEslintrc: true, baseConfig: {} }
+const ESLint = patch.eslint.ESLint
+const options = { extensions: [".js"], ignore: true, useEslintrc: true, baseConfig: {} }
```
Comment (fede-moya) [inline lib/eslint.js:15]: why is `reset: false` no longer present in `options` ?

File: lib/eslint.js [hunk 5]
```diff
@@ -77,7 +88,7 @@ function run(console, runOptions) {
           },
         },
       },
-      remediation_points: checks.remediationPoints(checkName, message, cli.getConfigForFile(path)),
+      remediation_points: checks.remediationPoints(checkName, message, cli.calculateConfigForFile(path)),
```
Comment (fede-moya) [inline lib/eslint.js:91]: I guess this is just a name change on eslint internals api, from `getConfigForFile` to `calculateConfigFile` ? Didn't see a note about that on eslint's release notes, can you explain why is this necessary ?

File: lib/eslint.js [hunk 6]
```diff
@@ -184,11 +197,12 @@ function run(console, runOptions) {
         console.error("Analyzing: " + batchFiles)
       }
 
-      runWithTiming("analyze-batch-" + batchNum, function() {
-        batchReport = cli.executeOnFiles(batchFiles)
+      await runWithTimingAsync("analyze-batch" + batchNum, async function() {
```
Comment (fede-moya) [inline lib/eslint.js:200]: I noticed you change `analyze-batch-` to just `analyze-batch` is this intentional ?

File: lib/eslint.js (hunk 5 above)
Comment (fede-moya) [reply lib/eslint.js:91]: Ok found the explanation here https://eslint.org/docs/user-guide/migrating-to-8.0.0#remove-cliengine

File: lib/eslint.js [hunk 7]
```diff
@@ -251,13 +265,13 @@ function run(console, runOptions) {
     }
   })
 
-  cli = new CLIEngine(options)
+  cli = new ESLint(options)
```
Comment (fede-moya) [inline lib/eslint.js:268]: this is part of one of eslint 8 changes, `CLIEngine` has been replace for `ESLint` as pointed on https://eslint.org/docs/user-guide/migrating-to-8.0.0#remove-cliengine

File: system-tests/complex-config/eslintrc.json [hunk 8]
```diff
@@ -1,6 +1,7 @@
 {
-  "parser": "babel-eslint",
+  "parser": "@babel/eslint-parser",
   "parserOptions": {
+    "requireConfigFile": false,
     "ecmaVersion": 2018,
     "sourceType": "module",
     "ecmaFeatures": {
```
Comment (fede-moya) [inline system-tests/complex-config/eslintrc.json:7]: could you elaborate on why this changes are needed to make the tests pass ?

File: system-tests/extends_airbnb/.eslintrc.js [hunk 9]
```diff
@@ -1,5 +1,8 @@
 module.exports = {
   extends: ["airbnb"],
-  parser: "babel-eslint",
+  parser: "@babel/eslint-parser",
+  parserOptions: {
+    requireConfigFile: false,
+  },
```
Comment (fede-moya) [inline system-tests/extends_airbnb/.eslintrc.js:6]: same as above, it's not clear to me why is this needed at least not after reading the eslint 8 migration document.

File: lib/eslint6-patch.js [hunk 10]
```diff
@@ -1,7 +1,11 @@
 "use strict"
 
-const { ConfigArrayFactory } = require("eslint/lib/cli-engine/config-array-factory")
-const { ConfigArray } = require("eslint/lib/cli-engine/config-array")
+const {
+  Legacy: {
+      ConfigArray,
+      ConfigArrayFactory
+  }
+} = require("@eslint/eslintrc")
```
Comment (fede-moya) [inline lib/eslint6-patch.js:8]: it may be worth it to invest some time on figuring how to use the new versions of ConfigArrayFactory and ConfigArray ? Also, what do you you think about renaming this file to `eslint8-patch.js` instead ?

File: integration/eslint_test.js [hunk 11]
```diff
@@ -57,7 +56,32 @@ describe("eslint integration", function() {
...
+      await fs.promises.writeFile(
+        configPath,
+        JSON.stringify({
+          enabled: true,
+          config: config,
+          include_paths: [sourcePath],
+        })
+      )
+      return await cb(directory)
+    }
```
Comment (fede-moya) [inline integration/eslint_test.js:84]: can you add a comment about what is being tested here 🙏🏼

Comment (fede-moya) [reply integration/eslint_test.js:84]: and thanks for adding the test

**Review by fede-moya — COMMENTED** [review]
So great that we are receiving eslint-8 🔥  🔥  - Thank you for taking  the initiative.

Left some comments. Also it would be great if you could include an overview of the non backwards compatible changes introduced in eslint-8 on the pr description. Or links to docs like https://github.com/eslint/eslint/pull/14934/files.

Tested locally and it works as expected 🚀 🚀 🚀 🚀 .

However, I noticed something weird when running `codeclimate --debug -e eslint`

```
D, [2022-02-09T19:49:19.469953 #1] DEBUG -- : engine stderr: Module not supported: @typescript-eslint
D, [2022-02-09T19:49:19.470394 #1] DEBUG -- : engine stderr: Module not supported: plugin:@typescript-eslint/eslint-recommended
D, [2022-02-09T19:49:19.476844 #1] DEBUG -- : engine stderr: Module not supported: plugin:@typescript-eslint/recommended
D, [2022-02-09T19:49:22.3
```

And I'm not seeing that when I use `channel/eslint-7` so maybe this work needs another PR to update the typescript plugin ?

File: .codeclimate.yml [hunk 12]
```diff
@@ -1,14 +1,14 @@
...
-    channel: eslint-6
+    channel: eslint-7
   duplication:
     enabled: true
     config:
       languages:
       - javascript
   shellcheck:
-    enabled: true
+    enabled: false
```
[the remark this answers is not in the document]
Comment (dantevvp) [reply .codeclimate.yml:11]: Whoops, disabled it to test something and forgot to turn it back on

File: lib/eslint.js (hunk 2 above)
Comment (dantevvp) [reply lib/eslint.js:167]: Yes, this is the new name of the `configFile` attribute in the [options object](https://github.com/eslint/eslint/blob/main/lib/eslint/eslint.js#L191).

File: lib/eslint.js (hunk 4 above)
Comment (dantevvp) [reply lib/eslint.js:15]: The `reset` option is [no longer available](https://eslint.org/docs/developer-guide/nodejs-api)

File: lib/eslint.js (hunk 6 above)
Comment (dantevvp) [reply lib/eslint.js:200]: Not intentional, fixing

File: .codeclimate.yml [hunk 13]
```diff
@@ -1,14 +1,14 @@
 engines:
   eslint:
     enabled: true
-    channel: eslint-6
+    channel: eslint-7
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply .codeclimate.yml:4]: haha 😮‍💨  you are 100% right

File: .eslintrc [hunk 14]
```diff
@@ -8,7 +8,7 @@
     "block-spacing": 2,
     "brace-style": [2, "1tbs", { "allowSingleLine": true }],
     "comma-dangle": [2, "never"],
-    "comma-style": [2, "first", { exceptions: {ArrayExpression: true, ObjectExpression: true} }],
+    "comma-style": [2, "first", { "exceptions": {"ArrayExpression": true, "ObjectExpression": true } }],
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply .eslintrc:11]: good call 👍🏼

File: package.json [hunk 15]
```diff
@@ -21,7 +21,8 @@
     "@typescript-eslint/parser": "^4.33.0",
     "babel-eslint": "^10.0.1",
     "babel-plugin-react-generate-property": "^1.0.5",
-    "eslint": "7.8.1",
+    "chai-as-promised": "^7.1.1",
```
[the remark this answers is not in the document]
Comment (fede-moya) [reply package.json]: thank u

File: lib/eslint6-patch.js (hunk 10 above)
Comment (fede-moya) [reply lib/eslint6-patch.js:8]: okay, that makes sense

**Review by fede-moya — APPROVED** [review]
This looks great Dante, good work 💪🏼 . Also thanks for being so diligent with the given feedback. I think the next move here is to deploy this to a `beta` channel and run some extensive test on production.
